django/db/models/base.py

```diff
@@ -803,18 +803,20 @@ def save(
                 RemovedInDjango60Warning,
                 stacklevel=2,
             )
-            for arg, attr in zip(
-                args, ["force_insert", "force_update", "using", "update_fields"]
-            ):
-                if arg:
-                    if attr == "force_insert":
-                        force_insert = arg
-                    elif attr == "force_update":
-                        force_update = arg
-                    elif attr == "using":
-                        using = arg
-                    else:
-                        update_fields = arg
+            force_insert = args[0]
+            len_args = len(args)
+            if len_args >= 2:
+                force_update = args[1]
+            if len_args >= 3:
+                using = args[2]
+            if len_args >= 4:
+                update_fields = args[3]
+            if len_args > 4:
+                # Recreate the message from Python
+                raise TypeError(
+                    "Model.save() takes from 1 to 5 positional arguments but "
+                    f"{len_args + 1} were given"
+                )
```

Comment:
Just thought of this shorter, maybe faster, maybe less clear alternative:

```python
if len(args) > 4:
                # Recreate the message from Python
                raise TypeError(
                    "Model.save() takes from 1 to 5 positional arguments but "
                    f"{len(args) + 1} were given"
                )
            force_insert = args[0]
            try:
                force_update = args[1]
                using = args[2]
                update_fields = args[3]
            except IndexError:
                pass
```

what do you think?

Reply:
My personal preference, and also in order to comply with our documented f-string format style (_f-strings should use only plain variable and property access_):
```python
            total_len_args = len(args) + 1  # include self
            if total_len_args > 5:
                # Recreate the proper TypeError message from Python.
                raise TypeError(
                    "Model.save() takes from 1 to 5 positional arguments but "
                    f"{total_len_args} were given"
                )
            force_insert = args[0]
            try:
                force_update = args[1]
                using = args[2]
                update_fields = args[3]
            except IndexError:
                pass
```
I believe a few more tests are needed because if we comment lines from the above, no test fails. Specifically, removing/commenting out all these lines:
```python
            # force_insert = args[0]
            try:
                # force_update = args[1]
                # using = args[2]
```
produces no test failures.

Reply:
Something like this would work as a test:
```python
    def test_save_deprecation_positional_arguments_used(self):
        a = Article()
        fields = ["headline"]
        with (
            self.assertWarns(RemovedInDjango60Warning),
            mock.patch.object(a, "save_base") as mock_save_base,
        ):
            a.save(None, 1, 2, fields)
        mock_save_base.assert_called_once_with(
            using=2, force_insert=None, force_update=1, update_fields=frozenset(fields))
```

Reply:
Thanks, that’s great. Added to my commit. I changed the test to assert on the `mock_calls` attribute, which I think is clearer and less error-prone, since there’s no chance to mistype `assert_called_onec_with`.